src/odb/src/db/dbModInst.cpp

```diff
@@ -405,6 +406,130 @@ void dbModInst::RemoveUnusedPortsAndPins()
+  dbSet<dbModBTerm> old_bterms = old_module->getModBTerms();
+  dbSet<dbModBTerm> new_bterms = new_module->getModBTerms();
+  if (old_bterms.size() != new_bterms.size()) {
+    logger->warn(utl::ODB,
+                 453,
+                 "modules cannot be swapped because module {} "
+                 "has {} ports but module {} has {} ports",
+                 old_module_name,
+                 old_bterms.size(),
+                 new_module_name,
+                 new_bterms.size());
+    return false;
+  }
+
+  // Check if module port names match
+  std::vector<_dbModBTerm*> new_ports;
+  std::vector<_dbModBTerm*> old_ports;
+  dbSet<dbModBTerm>::iterator iter;
+  for (iter = old_bterms.begin(); iter != old_bterms.end(); ++iter) {
+    old_ports.push_back((_dbModBTerm*) *iter);
+  }
+  for (iter = new_bterms.begin(); iter != new_bterms.end(); ++iter) {
+    new_ports.push_back((_dbModBTerm*) *iter);
+  }
+  std::map<dbModNet*, dbModNet*> mod_map;  // old mod net -> new mod net
+  std::sort(new_ports.begin(),
+            new_ports.end(),
+            [](_dbModBTerm* port1, _dbModBTerm* port2) {
+              return strcmp(port1->_name, port2->_name) < 0;
+            });
+  std::sort(old_ports.begin(),
+            old_ports.end(),
+            [](_dbModBTerm* port1, _dbModBTerm* port2) {
+              return strcmp(port1->_name, port2->_name) < 0;
+            });
+  std::vector<_dbModBTerm*>::iterator i1 = new_ports.begin();
+  std::vector<_dbModBTerm*>::iterator i2 = old_ports.begin();
+  for (; i1 != new_ports.end() && i2 != old_ports.end(); ++i1, ++i2) {
+    _dbModBTerm* t1 = *i1;
+    _dbModBTerm* t2 = *i2;
+    if (strcmp(t1->_name, t2->_name) != 0) {
+      break;
+    }
+    // Map old mod net to new mod net
+    mod_map[((dbModBTerm*) t2)->getModNet()] = ((dbModBTerm*) t1)->getModNet();
+  }
+  if (i1 != new_ports.end() || i2 != old_ports.end()) {
+    logger->warn(utl::ODB,
+                 454,
+                 "modules cannot be swapped because module {} "
+                 "has port {} but module {} has port {}",
+                 old_module_name,
+                 (*i1)->_name,
+                 new_module_name,
+                 (*i2)->_name);
+    return false;
+  }
+
+  _dbModule* new_master = (_dbModule*) dbModule::copy(new_module, this);
+
+  // Patch connections such that boundary nets connect to new module iterms
+  // instead of old module iterms
+  debugPrint(logger,
+             utl::ODB,
+             "replace_design",
+             1,
+             "Connecting nets that span module boundary");
+  for (const auto& [old_mod_net, new_mod_net] : mod_map) {
+    dbSet<dbITerm> old_iterms = old_mod_net->getITerms();
+    dbSet<dbITerm> new_iterms = new_mod_net->getITerms();
+    dbSet<dbITerm>::iterator it_iter;
+    for (it_iter = old_iterms.begin(); it_iter != old_iterms.end(); ++it_iter) {
+      dbITerm* old_iterm = *it_iter;
+      dbNet* flat_net = old_iterm->getNet();
+      old_iterm->disconnect();
```

Comment:
Note that this will disconnect both the modnet and dbNet connected to old_iterm. Perhaps it is appropriate to save the dbModNet* too. (but see comment later about copyModuleBoundaryIO).

Reply:
Got it. Fixed